### PR TITLE
[incubator/patroni] PostgreSQL 10 and best practices

### PR DESCRIPTION
**Pull request by hiddeco**
:warning: ~**https://github.com/kubernetes/charts/pull/3953 and https://github.com/kubernetes/charts/pull/3954 should go first**~ :warning:

Applies best practices, defaults to Patroni / Spilo releases based upon PostgreSQL 10 and re-implements admin password functionalities.

### CODE REVIEW

File: incubator/patroni/templates/ep-patroni.yaml
Comment (unguiculus) [inline incubator/patroni/templates/ep-patroni.yaml]: `app` labels should be `{{ template "patroni.name" . }}`

File: incubator/patroni/templates/ep-patroni.yaml
Comment (unguiculus) [inline incubator/patroni/templates/ep-patroni.yaml]: I'd suggest you remove the release name prefix. A component isn't tied to a release.

File: incubator/patroni/values.yaml
Comment (unguiculus) [inline incubator/patroni/values.yaml]: We recommend not to specify defaults for resources.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for this. I appreciate your effort. I made a few comments. Also I suggest you look at our review guidelines again for best practices. It'd also be good to add the current `_helpers.tpl` you get with `helm create` using Helm 2.8+.

Have you thought about updating the chart such that it can work without Etcd or Zookeeper? I've played around with this here when support for this was still experimental and not merged yet: https://github.com/unguiculus/charts/tree/feature/patroni/incubator/patroni. It's however a little outdated now.

**Comment by hiddeco** [comment]
@unguiculus agree with your comments and will push an update sometime this week.

Would be more than happy to add native Kubernetes support to the chart besides the etcd and ZooKeeper support it already offers. Reason I did not add this yet is that I need to take some time to look in to the functional details and I'd like to keep my PRs in nicely separated chunks of improvements.

I also have to add this has no priority for my employee as the native Kubernetes option has some drawbacks on GKE/GCE due to masters restarting in read-only modus when the Kubernetes API becomes unavailable (https://github.com/zalando/patroni/issues/574#issuecomment-353000821). But as you already paved the road for me in your branch I suspect it will take little of my time.

**Comment by hiddeco** [comment]
@unguiculus I have updated the PR with several improvements:

- Updated the `_helpers.tpl` to match Helm 2.8+
- Updated the labels to match Helm 2.8+
- Dropped the `component` label in favour of the (full)name override it now offers
- Refactored the `values.yaml` one more time:
    - image options are now under the `image` key, also moved the `imagePullPolicy` to `image.pullPolicy`
	- `replicas` -> `replicaCount`
	- Added/refactored `resources`, `nodeSelector`, `tolerations` and `affinity` options
- Added `selector` to the StatefulSet
- Revised the README

File: incubator/patroni/README.md
Comment (timstoop) [inline incubator/patroni/README.md]: While you're add it, I'd recommend making this `Kubernetes 1.5+`, as I assume it's been tested on newer versions as well.

File: incubator/patroni/Chart.yaml
Comment (timstoop) [inline incubator/patroni/Chart.yaml]: Please use the full 2-dot notation (0.7.0 in this case).

**Comment by timstoop** [comment]
Sorry, didn't mean to do that

**Review by timstoop — CHANGES_REQUESTED** [review]
One small change request, one small change suggestion.

**Comment by hiddeco** [comment]
@unguiculus waiting for your approval whenever you find time.

File: incubator/patroni/templates/statefulset-patroni.yaml
Comment (unguiculus) [inline incubator/patroni/templates/statefulset-patroni.yaml]: Please allow affinity to be configured the same way as e. g. node selectors or tolerations. This would be more flexible and would allow you to consider e. g. availability zones as well and not just hostnames.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/patroni/templates/statefulset-patroni.yaml
Comment (hiddeco) [reply incubator/patroni/templates/statefulset-patroni.yaml]: Added the option to set `affinity` rules, if you set these the `antiAffinity` will be ignored.

Comment (unguiculus) [reply incubator/patroni/templates/statefulset-patroni.yaml]: I don't like this. The `podAntiAffinity` block is rather confusing. You should remove it. You have the full flexibilty to configure anything you need. You still may provide a default as follows using the `tpl` function.

https://github.com/kubernetes/charts/blob/494b80b66acdef9de90bcaf77eb1e5298ff9c0fc/incubator/vault/templates/deployment.yaml#L54
https://github.com/kubernetes/charts/blob/494b80b66acdef9de90bcaf77eb1e5298ff9c0fc/incubator/vault/values.yaml#L40-L49

Comment (hiddeco) [reply incubator/patroni/templates/statefulset-patroni.yaml]: I wasn't aware of the `tpl` function, updated.

**Review by unguiculus — APPROVED** [review]

**Comment by hiddeco** [comment]
@unguiculus what is blocking the merge?

**Pull request merged**
